components/places/src/storage.rs

```diff
@@ -263,21 +270,311 @@ fn add_visit(db: &impl ConnExt,
+            Ok(Self {
+                url: Url::parse(&row.get_checked::<_, String>("url")?)?,
+                guid: SyncGuid(row.get_checked::<_, String>("guid")?),
+                row_id: row.get_checked("id")?,
+                title: row.get_checked::<_, Option<String>>("title")?.unwrap_or_default(),
+            })
+        }
+    }
+
+    pub fn fetch_visits(db: &Connection, url: &Url, limit: usize) -> Result<Option<(FetchedVisitPage, Vec<FetchedVisit>)>> {
+        // We do this in 2 steps - "do we have a page" then "get visits"
+        let page_sql = "
+          SELECT guid, url, id, title
+          FROM moz_places h
+          WHERE url_hash = hash(:url) AND url = :url";
+
+        let page_info = match db.try_query_row(page_sql, &[(":url", &url.to_string())], FetchedVisitPage::from_row, true)? {
+            None => return Ok(None),
+            Some(pi) => pi,
+        };
+
+        let mut stmt = db.prepare("
+          SELECT is_local, visit_type, visit_date
+          FROM moz_historyvisits
+          WHERE place_id = :place_id
+          LIMIT :limit")?;
+        let visits = stmt.query_and_then_named(&[(":place_id", &page_info.row_id),
+                                                 (":limit", &(limit as u32))],
+                                               FetchedVisit::from_row)?
+                         .collect::<Result<Vec<_>>>()?;
+        Ok(Some((page_info, visits)))
+    }
+
+    /// Apply history visit from sync. This assumes they have all been
+    /// validated, deduped, etc - it's just the storage we do here.
+    pub fn apply_synced_visits(db: &Connection,
+                               new_guid: &SyncGuid,
+                               existing_guid: &Option<SyncGuid>,
+                               url: &Url, title: &Option<String>,
+                               visits: &[HistoryRecordVisit]) -> Result<()> {
+        let page_info = match fetch_page_info(db, &url)? {
+            Some(info) => {
+                assert_eq!(info.page.guid, *new_guid);
+                info.page
+            },
+            None => {
+                assert!(existing_guid.is_none());
+                new_page_info(db, &url, Some(new_guid.clone()))?
+            },
+        };
+        for visit in visits {
+            let transition = VisitTransition::from_primitive(visit.transition).expect("these should already be validated");
+            add_visit(db, &page_info.row_id, &None, &visit.date.into(), &transition, &false)?;
+        }
+        // XXX - we really need a better story for frecency-boost than
+        // Option<bool> - None vs Some(false) is confusing. We should use an enum.
+        update_frecency(&db, page_info.row_id, None)?;
+
+        // and the place itself if necessary.
+        let new_title = title.as_ref().unwrap_or(&page_info.title);
+        // We set the Status to Normal, otherwise we will re-upload it as
+        // outgoing even if nothing has changed.
```

Comment:
I've left this as is for now, but added a comment indicating why we don't change the change counter (because a non-zero change counter before means there are changes we should upload before reconcilliation, so we should upload it after reconcilliation)

(Note that a test fails if we reset this to zero, which should help understand why)